middle_end/flambda/invariant_params.ml

```diff
@@ -65,47 +65,34 @@ let implies relation from to_ =
       relation
 
 let transitive_closure state =
-  let union s1 s2 =
-    match s1, s2 with
-    | Top, _ | _, Top -> Top
-    | Implication s1, Implication s2 ->
-      Implication (Variable.Pair.Set.union s1 s2)
+  (* Depth-first search for all implications for one argument.
+     Arguments are moved from candidate to frontier, assuming
+     they are newly added to the result. *)
+  let rec loop candidate frontier result =
+    match (candidate, frontier) with
+    | ([], []) -> Implication result
+    | ([], frontier::fs) ->
+      (* Obtain fresh candidate for the frontier argument. *)
+      (match (try Variable.Pair.Map.find frontier state with
+              | Not_found -> Implication Variable.Pair.Set.empty) with
+       | Top -> Top
+       | Implication candidate ->
+         loop (Variable.Pair.Set.elements candidate) fs result)
+    | (candidate::cs, frontier) ->
+      let result' = Variable.Pair.Set.add candidate result in
+      if result' != result then
+        (* Result change means candidate becomes part of frontier. *)
+        loop cs (candidate :: frontier) result'
+      else
+        loop cs frontier result
   in
-  let equal s1 s2 =
-    match s1, s2 with
-    | Top, Implication _ | Implication _, Top -> false
-    | Top, Top -> true
-    | Implication s1, Implication s2 -> Variable.Pair.Set.equal s1 s2
-  in
-  let update arg state =
-    let original_set =
-      try Variable.Pair.Map.find arg state with
-      | Not_found -> Implication Variable.Pair.Set.empty
-    in
-    match original_set with
-    | Top -> state
-    | Implication arguments ->
-        let set =
-          Variable.Pair.Set.fold
-            (fun orig acc->
-               let set =
-                 try Variable.Pair.Map.find orig state with
-                 | Not_found -> Implication Variable.Pair.Set.empty in
-               union set acc)
-            arguments original_set
-        in
-        Variable.Pair.Map.add arg set state
-  in
-  let once state =
-    Variable.Pair.Map.fold (fun arg _ state -> update arg state) state state
-  in
-  let rec fp state =
-    let state' = once state in
-    if Variable.Pair.Map.equal equal state state'
-    then state
-    else fp state'
-  in
-  fp state
+    Variable.Pair.Map.filter_map
+      (fun _ set ->
+         Some (match set with
+	       | Top -> Top
+	       | Implication set ->
+	         loop [] (Variable.Pair.Set.elements set) set))
```

Comment:
If I understand correctly, you are computing the transitive closure of each element of the input state independently -- without reusing the computations you have already done for the previous elements. This sounds wasteful: if `(v1, s1)` is in the input state and we compute its transitive closure `s1'`, and then we encounter another element `(v2, s2)` and `v1` belongs to `s2` (or is found later when computing the closure from `s2`), iiuc. you will recompute `s1'` from `s1` at this point whereas you could directly use `s1'`. I would thus rather expect a fold on the input state rather than a map. Am I missing something?

Reply:
(In fact it might even be possible to compute the transitive closure for the whole mapping at once, instead of computing element-per-element, which saves even more work as intermediate computations can also be reused. I suppose that the "candidates" would be a pair `variable * candidate list`, and the frontier would be a list `(variable * candidate list) list`. But this is a slightly more invasive change than just using a `fold` above, and maybe there is no performance justification for making the code slightly more complex.)

Reply:
For a non-expert trying to understand the code (I haven't studied the details as you have, so apologies for any misunderstandings), the problem with the previous implementation is the fact that whenever any part of the state changes, all variables are recomputed on the next iteration, leading to wasteful recomputations that are probably quadratic in practice.

You are pointing out that with my suggestion there is a risk of unnecessary work as well: if `s1'` is much larger than `s1`, but most of it is already in `s2`, it is wasteful to compute the union `s1' U s2` rather than `s1 U s2`. My intuition would be that this is fine as the union computation is efficient when sets overlap.

> With the fold, it's no longer obvious that the individual steps for each node process at most m relations (m being the number of input relations) because we get extra relations from previous steps.

Yes, but on the other hand the number of steps may increase a lot with the `map` due to duplicated closure work -- in the worst case of a fully connected graph, the computation goes from `O(n)` to `O(n^2)`.

Reply:
Could you try the small tweak of making `frontier` a set instead of a list?